package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,943 @@
+            self.nbeads[i] = 0
+
+    def __dealloc__(self):
+        PyMem_Free(self.nbeads)
+        PyMem_Free(self.beadids)
+        PyMem_Free(self.cellids)
+
+    cdef ns_int coord2cellid(self, rvec coord) nogil:
+        return <ns_int> (coord[ZZ] / self.cellsize[ZZ]) * (self.ncells[XX] * self.ncells[YY]) +\
+               <ns_int> (coord[YY] / self.cellsize[YY]) * self.ncells[XX] + \
+               <ns_int> (coord[XX] / self.cellsize[XX])
+
+    cdef bint cellid2cellxyz(self, ns_int cellid, ivec cellxyz) nogil:
+        if cellid < 0:
+            return False
+        if cellid >= self.size:
+            return False
+
+        cellxyz[ZZ] = <ns_int> (cellid / self.cell_offsets[ZZ])
+        cellid -= cellxyz[ZZ] * self.cell_offsets[ZZ]
+
+        cellxyz[YY] = <ns_int> (cellid / self.cell_offsets[YY])
+        cellxyz[XX] = cellid - cellxyz[YY] * self.cell_offsets[YY]
+
+        return True
+
+    cdef fill_grid(self, real[:, ::1] coords):
+        cdef ns_int i, cellindex = -1
+        cdef ns_int ncoords = coords.shape[0]
+        cdef ns_int *beadcounts = NULL
+
+        # Allocate memory
+        beadcounts = <ns_int *> PyMem_Malloc(sizeof(ns_int) * self.size)
+        if not beadcounts:
+            raise MemoryError("Could not allocate memory for bead count buffer ({} bits requested)".format(sizeof(ns_int) * self.size))
+
+        with nogil:
+            # Initialize buffers
+            for i in range(self.size):
+                beadcounts[i] = 0
+
+            # First loop: find cellindex for each bead
+            for i in range(ncoords):
+                cellindex = self.coord2cellid(&coords[i, 0])
+
+                self.nbeads[cellindex] += 1
+                self.cellids[i] = cellindex
+
+                if self.nbeads[cellindex] > self.nbeads_per_cell:
+                    self.nbeads_per_cell = self.nbeads[cellindex]
+
+            # Allocate memory
+            with gil:
+                self.beadids = <ns_int *> PyMem_Malloc(sizeof(ns_int) * self.size * self.nbeads_per_cell) #np.empty((self.size, nbeads_max), dtype=np.int)
+                if not self.beadids:
+                    raise MemoryError("Could not allocate memory for NSGrid.beadids ({} bits requested)".format(sizeof(ns_int) * self.size * self.nbeads_per_cell))
+
+            # Second loop: fill grid
+            for i in range(ncoords):
+
+                # Add bead to grid cell
+                cellindex = self.cellids[i]
+                self.beadids[cellindex * self.nbeads_per_cell + beadcounts[cellindex]] = i
+                beadcounts[cellindex] += 1
+
+        # Now we can free the allocation buffer
+        PyMem_Free(beadcounts)
+
+
+cdef class FastNS(object):
+    cdef bint debug
+    cdef PBCBox box
+    cdef real[:, ::1] coords
+    cdef real[:, ::1] coords_bbox
+    cdef readonly real cutoff
+    cdef bint prepared
+    cdef NSGrid grid
+
+    def __init__(self, u, cutoff, coords=None, prepare=True, debug=False, max_gridsize=5000):

```

Comment:
it doesn't need a universe. Just use a `coords` and `box` argument like the other functions in lib

Reply:
As pointed out by @kain88-de, the algorithm doest not need a `Universe` and it can work with just an array of coords and the box vectors. It really depends on how the NS is considered:
1. it could be used to guess bonds when topology is not available (eg .gro files). In which case, it could become a core feature
2. it stays a "battery included" with MDAnalysis. In which case, the use of `Universe`is questionnable

I must admit I changed the original code to use `Universe` instance with case 1 in mind and to ease the call for end-users.

Reply:
Having it work without a universe would be nice. I do have occasionally some use-cases where I need such an algorithm and only have raw coordinates and box dimensions.

Reply:
I'd put the bare bones (no `Universe`) code into `lib.nsgrid` and then we can always add a thin wrapper somewhere else. For right now you could just add an example of how to use it with a `Universe`.